tests/Highlighter.py

```diff
@@ -0,0 +1,303 @@
+                    tok = current_rule['token'][0]
+                    start = current_match.start()
+                    length = current_match.end() - current_match.start()
+                    output = output[:start] + TOKEN_CODE[tok] * length + output[start + length:]
+
+                else:
+                    pos = current_match.start()
+                    for i,submatch in enumerate(current_match.groups()):    
+                        tok = current_rule['token'][i%len(current_rule['token'])]        
+                        output = output[:pos] + TOKEN_CODE[tok] * len(submatch) + output[pos + len(submatch):]
+                        pos += len(submatch)
+
+                current_position = current_match.end()
+
+                if 'next' in current_rule:
+                    current_state = current_rule['next']
+
+                if current_position ==len(code):
+                    flag = True
+
+
+            else:
+                flag = True
+
+        output = output.replace(" ","T")
+        return output,current_state
+
+
+    def apply_rules(self,rules,code):
+        outputs = []
+        token = "start"
+        for line in code.split("\n"):
+            output,token = self.apply_rules_line(rules,line,token)
+            print("next",token)
+            outputs.append(output)
+        return "\n".join(outputs)
+
+
+
+    # This function allows to check 2 syntactic colorations one desired
+    # and the other obtained, and manages some special cases
+
+    # Returns a boolean and a value:
+    # - If the 2 highlights are consistent, the boolean will be true and the value unused
+    # - if there is an error in the highlight, the boolean will be false and
+    #   the value will be the number of the 1st character where there is an inconsistency
+    def check(self,result,expected):
+        if len(result) != len(expected):
+            raise ValueError("The desired highlight and the obtained highlight do not have the same length !")
+        cpt = 0
+        for i in range(len(expected)):
+            ch_wanted,ch_result = expected[i],result[i]
+
+            if ch_wanted == ch_result:
+                pass # if they are the same, no problem 
+            elif ch_wanted == " " and ch_result in ['T','K','C','N','S']:
+                pass # if it was not fixed at the beginning, everything is tolerated except the pink highlighting
+            else:
+                # in all other cases, an error is returned indicating the number of the problematic character
+                return False,cpt
+
+            cpt += 1
+            if result[i] == "\n" : cpt += 1
+
+        return True,-1
+
+
+
+    #######################################################################################
+    #######################################################################################
+    #######################################################################################
```

Comment:
This looks pretty but doesn't add any value (except set your code apart from other code, which we agreed is a bad thing)